### PR TITLE
[core] Adds PinRuntimeEnvUri API to GcsClient.

### PR DESCRIPTION
**Pull request by rynewang**
This is to bring feature parity for GcsClient with PythonGcsClient. Like https://github.com/ray-project/ray/pull/46061

### CODE REVIEW

File: src/ray/gcs/gcs_client/accessor.cc [hunk 1]
```diff
@@ -1294,5 +1294,20 @@ Status InternalKVAccessor::Exists(const std::string &ns,
...
+Status RuntimeEnvAccessor::PinRuntimeEnvUri(const std::string &uri,
+                                            int expiration_s,
+                                            int64_t timeout_ms) {
+  rpc::PinRuntimeEnvURIRequest request;
+  request.set_uri(uri);
+  request.set_expiration_s(expiration_s);
+  rpc::PinRuntimeEnvURIReply reply;
+  auto status =
+      client_impl_->GetGcsRpcClient().SyncPinRuntimeEnvURI(request, &reply, timeout_ms);
+  return status;
```
Comment (ruisearch42) [inline src/ray/gcs/gcs_client/accessor.cc:1309]: Any reason we skipped checking and updating `status` as in `PythonGcsClient::PinRuntimeEnvUri()`?

File: src/ray/gcs/gcs_client/accessor.h [hunk 2]
```diff
@@ -885,6 +885,23 @@ class InternalKVAccessor {
...
+  RuntimeEnvAccessor() = default;
+  explicit RuntimeEnvAccessor(GcsClient *client_impl);
+  virtual ~RuntimeEnvAccessor() = default;
+
+  /// Pins a runtime environment by URI.
+  ///
+  /// Only works if URI has prefix "gcs://", for which GCS holds a reference for
+  /// `expiration_s` seconds. After that, GCS decrements the reference count.
+  ///
+  /// For all other URIs, this call is a no-op.
```
Comment (ruisearch42) [inline src/ray/gcs/gcs_client/accessor.h:899]: What would be the expected return value in this case? Is it considered as invalid arg, and do we want to return something like `Invalid` or `KeyError`? Could you please add doc for return value as well as params?

File: src/ray/gcs/gcs_client/accessor.cc [hunk 3]
```diff
@@ -1294,5 +1294,20 @@ Status InternalKVAccessor::Exists(const std::string &ns,
   return ret_promise.get_future().get();
 }
 
+RuntimeEnvAccessor::RuntimeEnvAccessor(GcsClient *client_impl)
+    : client_impl_(client_impl) {}
+
+Status RuntimeEnvAccessor::PinRuntimeEnvUri(const std::string &uri,
```
Comment (ruisearch42) [inline src/ray/gcs/gcs_client/accessor.cc:1300]: Can we add a test to verify the behavior is the same as `PythonGcsClient::PinRuntimeEnvUri()`, or maybe just confirm the expected changed behavior?

File: src/ray/gcs/gcs_client/accessor.h [hunk 4]
```diff
@@ -885,6 +885,23 @@ class InternalKVAccessor {
   GcsClient *client_impl_;
 };
 
+class RuntimeEnvAccessor {
+ public:
+  RuntimeEnvAccessor() = default;
```
Comment (ruisearch42) [inline src/ray/gcs/gcs_client/accessor.h:890]: wondering why we need this

File: src/ray/gcs/gcs_client/accessor.h (hunk 2 above)
Comment (rynewang) [reply src/ray/gcs/gcs_client/accessor.h:899]: We don't return anything error. The idea is, GCS "accepts" all kinds of rt envs, but only do actions on that one it recognizes. The whole design is a bit awkward I admit, thinking some solutions to remove it at all for good.

File: src/ray/gcs/gcs_client/accessor.cc (hunk 1 above)
Comment (rynewang) [reply src/ray/gcs/gcs_client/accessor.cc:1309]: GcsRpcClient method already checks the `status` within the proto reply: https://github.com/ray-project/ray/blob/master/src/ray/rpc/gcs_server/gcs_rpc_client.h#L211

File: src/ray/gcs/gcs_client/accessor.cc (hunk 3 above)
Comment (rynewang) [reply src/ray/gcs/gcs_client/accessor.cc:1300]: Behavior changes:

1. for GrpcUnknown and GrpcUnavailable, previously we fail, now we retry
2. No custom messages

File: src/ray/gcs/gcs_client/accessor.h (hunk 4 above)
Comment (rynewang) [reply src/ray/gcs/gcs_client/accessor.h:890]: All other accessors have it. I think it's for mocks: https://github.com/ray-project/ray/blob/ae3e87cb0bb51fecde718f1e9fbc4ced488c8986/src/mock/ray/gcs/gcs_client/accessor.h#L232

File: src/ray/gcs/gcs_client/accessor.h (hunk 2 above)
Comment (ruisearch42) [reply src/ray/gcs/gcs_client/accessor.h:899]: If it returns Status::OK for no-op, can we document it?

**Review by ruisearch42 — APPROVED** [review]

**Review by jjyao — APPROVED** [review]

**Pull request merged**